### PR TITLE
Create new plugin version list endpoint

### PR DESCRIPTION
**Pull request by crysxd**
#### What does this PR do and why is it necessary?
This PR introduces a new permission `LIST_PLUGINS` and a new endpoint `/plugin/pluginmanager/plugins/versions` which returns a compact list of all enabled plugins and their versions. This can be used by 3rd party systems that want to provide similar functionality to the web interface or integrate with plugins via their API. The new endpoint requires the `LIST_PLUGINS` permission which is grated by default to read only users, operators and admins. Guests do not have access to this endpoint (403). 

At the moment, only users holding the `MANAGE_PLUGIN` permission have the ability to see which plugins are installed. Other users can already "probe" for installed plugins by e.g. looking at the settings or trying to request API endpoints provided by plugins. This means the information is already available to malicious parties and the new endpoint does not expose additional information.

I opted to return a very compact data set as this is all that an external client would need. Additional information like e.g. safe mode victim, pending disabled etc. is not included as this data is only needed for the intend of managing the plugins - thus should be restricted to users holding the matching permission.

The endpoint returns a JSON object:

```
{
    "PrintTimeGenius": "2.2.8",
    "action_command_prompt": null,
    "announcements": null,
    "appkeys": null,
    "backup": null,
    "corewizard": null,
    "discovery": null,
    "errortracking": null,
    "eventmanager": null,
    "file_check": "2021.2.23",
    "firmware_check": "2021.10.11",
    "gcodeviewer": null,
    "logging": null,
    "resource_monitor": "0.3.5",
    "softwareupdate": null,
    "tracking": null,
    "virtual_printer": null
}
```

#### How was it tested? How can it be tested by the reviewer?
Postman & web interface

#### Any background context you want to provide?
[Discord discussion](https://discord.com/channels/704958479194128507/705047010641838211/914807730982772777)

I know we said we wouldn't need to filter for enabled plugins only, but I wanted to make the response as simple as possible. Now all listed plugins are available to be used. Hope that's ok, happy to change the structure if requested!

#### What are the relevant tickets if any?
/ 

#### 
Screenshots (if appropriate)
Screenshot 2021-12-11 at 15 09 59

#### Further notes
/

### CODE REVIEW

File: src/octoprint/plugins/pluginmanager/__init__.py
Comment (foosel) [inline src/octoprint/plugins/pluginmanager/__init__.py:533]: This will behave non-deterministically depending on the used python versions, because the order of items returned from a dict is not guaranteed to match insertion order in Python < 3.7.

Better:

``` python
hash_update(repr(list(map(lambda k, v: "{}={}".format(k, v), self._plugins_versions_response()))))
```

Alternatively you could also use `OrderedDict` for building the data structure.

Comment (foosel) [inline src/octoprint/plugins/pluginmanager/__init__.py:533]: This will behave non-deterministically depending on the used python versions, because the order of items returned from a dict is not guaranteed to match insertion order in Python < 3.7.

Better:

``` python
hash_update(repr(list(map(lambda k, v: "{}={}".format(k, v), self._plugins_versions_response()))))
```

Alternatively you could also use `OrderedDict` for building the data structure.

File: src/octoprint/plugins/pluginmanager/__init__.py
Comment (foosel) [inline src/octoprint/plugins/pluginmanager/__init__.py:477]:
```suggestion
    def _get_plugins_with_versions(self, only_enabled=True):
        return dict((p["key"], p["version"]) for p in self._get_plugins() if p["enabled"] or not only_enabled)
```

File: src/octoprint/plugins/pluginmanager/__init__.py
Comment (foosel) [inline src/octoprint/plugins/pluginmanager/__init__.py:519]: I think this doesn't belong here.

File: src/octoprint/plugins/pluginmanager/__init__.py
Comment (foosel) [inline src/octoprint/plugins/pluginmanager/__init__.py:540]: Considering that the full response is required in order to determine the ETag needed for revalidation checking, I wouldn't bother with caching here since it simply doesn't add anything in terms of saving resources.

If I need to fetch all enabled plugins & versions to figure out if they changed in order to respond with a `Not Modified` header, there's not much point in caching to begin with.

I'd rather suggest to get rid of the additional response generating method defined above altogether and inlining that here.

**Review by foosel — CHANGES_REQUESTED** [review]
Found a couple of points that need to be taken care of.

Quick heads-up, both a booster shot and a christmas break are coming up on my end, so it might take a bit until this gets merged depending on when changes are made.

**Comment by crysxd** [comment]
Will work through the changes today or tomorrow! Thanks for taking the time to check this out 
🙃 No time pressure here...so if we won't merge it before the new year that's totally fine...even it won't make it in 1.8

File: src/octoprint/plugins/pluginmanager/__init__.py
Comment (crysxd) [reply src/octoprint/plugins/pluginmanager/__init__.py:540]: So basically we can get rid of the entire ETag/caching and just return a simple json response?

Comment (foosel) [reply src/octoprint/plugins/pluginmanager/__init__.py:540]: IMHO yes. I mean, we don't really gain any performance from caching here, and in worst case (cache is stale) will actually have to do the work needed to preparing the response twice (once for the ETag comparison and once for the actual response generation).

Comment (crysxd) [reply src/octoprint/plugins/pluginmanager/__init__.py:540]: Totally agree...I basically just copied it from the other endpoints but guess the situation there is a bit different. Will remove it, also passively resolves some of the other points!

**Comment by crysxd** [comment]
Removed Etag as discussed and inlined the data generation

**Comment by crysxd** [comment]
Seems like npm has a bad day...I tried building it twice but installing eslint fails with 503

Will trigger it again later today

**Comment by foosel** [comment]
Hm, GitHub is still showing me the old code... did something go wrong with the force pushes there?

**Comment by crysxd** [comment]
Seems like I did something wrong when "restarting" the build this morning....sorry for the extra round :(

**Review by foosel — APPROVED** [review]

**Pull request merged**
